packages/flutter/lib/src/material/material_state.dart

```diff
@@ -66,6 +68,163 @@ enum MaterialState {
+  ///
+  /// The [builder], [disabled], [selected], and [error] arguments must be
+  /// non-null.
+  const MaterialStateBuilder({
+    Key key,
+    @required this.builder,
+    this.hovered,
+    this.focused,
+    this.pressed,
+    this.disabled = false,
+    this.selected = false,
+    this.error = false,
+  }) : assert(builder != null),
+       assert(disabled != null),
+       assert(selected != null),
+       assert(error != null),
+       super(key: key);
+
+  /// The callback that returns the child of this widget based on the current
+  /// `Set<MaterialState>`.
+  final Function(BuildContext context, Set<MaterialState> states) builder;
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.hovered].
+  ///
+  /// If `null`, widget will be wrapped in a [MouseRegion] to determine when
+  /// the child widget is being hovered over.
+  final bool hovered;
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.focused].
+  ///
+  /// If `null`, widget will be wrapped in a [Focus] to determine when the child
+  /// widget is in focus.
+  final bool focused;
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.pressed].
+  ///
+  /// If `null`, widget will be wrapped in a [Focus] to determine when the child
+  /// widget is in focus.
+  final bool pressed;
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.disabled].
+  ///
+  /// Defaults to `false`.
+  final bool disabled;
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.selected].
+  ///
+  /// Defaults to `false`.
+  final bool selected;
+
+  /// Whether or not the child of this widget is considered to be
+  /// [MaterialState.error].
+  ///
+  /// Defaults to `false`.
+  final bool error;
+
+  @override
+  _MaterialStateBuilderState createState() => _MaterialStateBuilderState();
+}
+
+class _MaterialStateBuilderState extends State<MaterialStateBuilder> {
+  final Set<MaterialState> _states = <MaterialState>{};
+
+  void _updateState(MaterialState state, bool value) {
+    if (_states.contains(state) != value) {
+      setState(() {
+        value ? _states.add(state) : _states.remove(state);
+      });
+    }
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    if (widget.hovered != null)
+      _updateState(MaterialState.hovered, widget.hovered);
+    if (widget.focused != null)
+      _updateState(MaterialState.focused, widget.focused);
+    if (widget.pressed != null)
+      _updateState(MaterialState.pressed, widget.pressed);
+    _updateState(MaterialState.disabled, widget.disabled);
+    _updateState(MaterialState.selected, widget.selected);
+    _updateState(MaterialState.error, widget.error);
+  }
+
+  @override
+  void didUpdateWidget(MaterialStateBuilder oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.hovered != null)
```

Comment:
All of this function is the same as `initState`, perhaps consolidate it to a common private function they both call?

Reply:
Done.